internal/acs/runner.go

```diff
@@ -0,0 +1,250 @@
+package acs
+
+import (
+	"fmt"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/app-sre/acs-cve-prom-exporter/internal/collector"
+	"github.com/app-sre/acs-cve-prom-exporter/pkg/acsclient"
+	"github.com/app-sre/acs-cve-prom-exporter/pkg/worker"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type CollectorLoopParams struct {
+	Collector      *collector.CVECollector
+	Logger         log.Logger
+	Credentials    ACSCredentials
+	ExitSignal     chan struct{}
+	WorkerPoolSize int
+	Timeout        int
+	Retry          int
+}
+
+type ACSCredentials struct {
+	Endpoint string
+	Token    string
+}
+type cveSummaryData struct {
+	clusterName     string
+	namespaceName   string
+	instances       map[string]int
+	imageID         string
+	imageRegistry   string
+	imageRemote     string
+	imageTag        string
+	operatingSystem string
+	cveData         map[string]cveData
+}
+type cveData struct {
+	severity  string
+	cvss      float64
+	isFixable bool
+}
+
+func CVECollectorLoop(collectorParams CollectorLoopParams) {
+	acsClient, err := acsclient.NewACSClient(collectorParams.Credentials.Endpoint, collectorParams.Credentials.Token, collectorParams.Timeout, collectorParams.Retry)
+	if err != nil {
+		level.Error(collectorParams.Logger).Log("msg", "error creating acs client", "error", err.Error())
+		collectorParams.ExitSignal <- struct{}{}
+		return
+	}
+	for {
+		cveData, err := getCVEDataFromACS(acsClient, collectorParams.WorkerPoolSize)
+		if err != nil {
+			level.Error(collectorParams.Logger).Log("msg", "Error processing metrics::", "error", err.Error())
+			collectorParams.ExitSignal <- struct{}{}
+			return
+		}
+
+		collectorParams.Collector.MetricsData = getMetricDataFromCVE(cveData, collectorParams.Logger, collectorParams.Collector)
+		level.Debug(collectorParams.Logger).Log("Collection completed.")
+	}
+}
+
+func getCVEDataFromACS(client acsclient.Client, workerPoolSize int) ([]interface{}, error) {
+	clusters, err := client.GetClusters()
+	if err != nil {
+		return nil, fmt.Errorf("error getting clusters:%v", err)
+	}
+	workerPool := worker.NewWorkerPool(workerPoolSize, true)
+	for _, cluster := range clusters {
+		// to avoid goroutine in loop error
+		clusterData := cluster
+		workerPool.Run(func() (interface{}, error) {
+			return processClusterCVEs(clusterData, client)
+		})
+	}
+	return workerPool.WaitAndReturnData()
+}
+
+func getMetricDataFromCVE(records []interface{}, logger log.Logger, collector *collector.CVECollector) map[string]prometheus.Metric {
```

Comment:
can `records` use a type instead of `interface{}` here?

Reply:
records is a slice because it consists the data from all the workers. at the end we append() to result..hence slice..i have tried to memic
`data = threaded.run(<func>)`

Reply:
we expect `records` to be a list of `map[string]cveSummaryData`, right? why need to type check inside this function instead of make it explicit in signature, any other types for `record`?

Reply:
we can use `map[string]cveSummaryData`...the reason to choose `[]interface{}` was to make `workerpools` generic and plug able at any stage to pull in data